File: tests/metrics/conditional_drawdown_at_risk_test.py
from __future__ import annotations
import numpy as np
import pytest
from taflow.metrics.conditional_drawdown_at_risk import ConditionalDrawdownAtRisk
PERFORMANCE_ANALYTICS_2_1_0_SHA256 = 'fc801d39382818cd3a7052326b45d078302aef4d290c85dab83498ed4516d58d'
RISKFOLIO_LIB_7_3_0_SHA256 = '73ab9400691e0ca4258ba1b3d4939018c67809955504807b0559904405246bdf'

def _performanceanalytics_discrete(returns: np.ndarray, confidence: float) -> float | None:
    """Independent translation of pinned CDD/findDrawdowns source."""
    usable = returns[~np.isnan(returns)]
    if len(usable) == 0:
        return None
    wealth = np.cumprod(1.0 + usable)
    peaks = np.maximum.accumulate(np.r_[1.0, wealth])[1:]
    drawdowns = wealth / peaks - 1.0
    episodes: list[float] = []
    prior_negative = bool(drawdowns[0] < 0.0)
    trough = float(drawdowns[0])
    for drawdown in drawdowns[1:]:
        negative = bool(drawdown < 0.0)
        if negative == prior_negative:
            trough = min(trough, float(drawdown))
        else:
            episodes.append(trough)
            prior_negative = negative
            trough = float(drawdown)
    episodes.append(trough)
    boundary = float(np.quantile(episodes, 1.0 - confidence, method='linear'))
    tail = np.asarray(episodes)[np.asarray(episodes) <= boundary]
    return -float(np.mean(tail))

@pytest.mark.parametrize('confidence', [0.5, 0.75, 0.9, 0.95, 0.99])
@pytest.mark.parametrize('returns', [np.array([-0.2]), np.array([0.2]), np.zeros(32), np.array([0.25, -0.2, 0.25, 0.1, -0.5, 1.0, -0.1]), np.array([0.25, -0.2, 0.25, 0.25, -0.2]), np.linspace(-0.04, 0.05, 101), np.array([np.nan, 0.02, -0.03, np.nan, 0.01]), np.random.default_rng(20260812).normal(0.0004, 0.012, 513)])
def test_matches_pinned_performanceanalytics_discrete_source(returns: np.ndarray, confidence: float) -> None:
    assert len(PERFORMANCE_ANALYTICS_2_1_0_SHA256) == 64
    assert len(RISKFOLIO_LIB_7_3_0_SHA256) == 64
    actual = ConditionalDrawdownAtRisk(confidence=confidence).from_returns(returns).compute()
    expected = _performanceanalytics_discrete(returns, confidence)
    assert actual == pytest.approx(expected, rel=1e-12, abs=1e-14)

def test_input_methods_and_lifecycle_are_invariant() -> None:
    returns = np.array([0.25, -0.2, 0.25, 0.1, -0.5, 1.0, -0.1])
    expected = ConditionalDrawdownAtRisk(confidence=0.75).from_returns(returns).compute()
    equity = 100.0 * np.r_[1.0, np.cumprod(1.0 + returns)]
    pnl = np.diff(equity)
    assert ConditionalDrawdownAtRisk(confidence=0.75).from_equity(equity).compute() == pytest.approx(expected)
    assert ConditionalDrawdownAtRisk(confidence=0.75).from_pnl(pnl, initial_capital=100.0).compute() == pytest.approx(expected)
    assert ConditionalDrawdownAtRisk(confidence=0.75).from_log_returns(np.log1p(returns)).compute() == pytest.approx(expected)
    state = ConditionalDrawdownAtRisk(confidence=0.75).from_returns([])
    assert state.value is None
    assert state.append(returns[0]) is state
    assert state.extend(returns[1:3]) is state
    assert state.extend(returns[3:]) is state
    assert state.compute() == pytest.approx(expected)
    assert state.compute() == pytest.approx(expected)
    assert len(state) == len(returns)
    assert state.reset() is state
    assert len(state) == 0
    assert state.value is None
    assert state.extend(returns).compute() == pytest.approx(expected)

def test_current_episode_zero_paths_and_boundary_ties() -> None:
    state = ConditionalDrawdownAtRisk().from_returns([0.25, -0.2])
    assert state.compute() == pytest.approx(0.2)
    assert state.append(0.25).compute() == pytest.approx(0.2)
    assert ConditionalDrawdownAtRisk().from_returns([0.1, 0.0, 0.2]).compute() == 0.0
    assert ConditionalDrawdownAtRisk().from_returns([0.25, -0.2, 0.25, 0.25, -0.2]).compute() == pytest.approx(0.2)

def test_missing_and_invalid_values_are_enforced() -> None:
    returns = np.array([0.25, np.nan, -0.2, 0.25, -0.1])
    usable = returns[~np.isnan(returns)]
    state = ConditionalDrawdownAtRisk().from_returns(returns)
    assert len(state) == len(usable)
    assert state.compute() == pytest.approx(_performanceanalytics_discrete(usable, 0.95))
    for confidence in [0.0, 1.0, -0.1, 1.1, np.nan, np.inf]:
        with pytest.raises(ValueError):
            ConditionalDrawdownAtRisk(confidence=confidence).from_returns([0.01])
    with pytest.raises(ValueError):
        ConditionalDrawdownAtRisk(nan_policy='raise').from_returns([np.nan])
    with pytest.raises(ValueError):
        ConditionalDrawdownAtRisk().from_returns([np.inf])
    with pytest.raises(ValueError):
        ConditionalDrawdownAtRisk().from_returns([-1.01])

def test_requires_semantic_input_method() -> None:
    metric = ConditionalDrawdownAtRisk()
    with pytest.raises(ValueError):
        metric.append(0.01)

File: crates/taflow-metrics/src/metrics/conditional_drawdown_at_risk_test.rs
use super::conditional_drawdown_at_risk::ConditionalDrawdownAtRisk;
use crate::{MetricInputKind, NanPolicy};
use approx::assert_relative_eq;

// PerformanceAnalytics 2.1.0 CRAN source tarball SHA-256:
// fc801d39382818cd3a7052326b45d078302aef4d290c85dab83498ed4516d58d.
// Its default discrete method applies R's type-7 quantile to the signed
// findDrawdowns episode troughs, selects every trough <= that boundary, and
// negates their mean. Riskfolio-Lib 7.3.0 CDaR_Rel is deliberately not this
// oracle: it uses every per-observation drawdown and fractional tail weighting.
fn pinned_performanceanalytics_discrete(returns: &[f64], confidence: f64) -> Option<f64> {
    if returns.is_empty() {
        return None;
    }
    let mut wealth = 1.0_f64;
    let mut peak = 1.0_f64;
    let mut episodes = Vec::new();
    let mut current: Option<(bool, f64)> = None;
    for &simple_return in returns {
        wealth *= 1.0 + simple_return;
        peak = peak.max(wealth);
        let drawdown = wealth / peak - 1.0;
        let negative = drawdown < 0.0;
        match current {
            None => current = Some((negative, drawdown)),
            Some((sign, trough)) if sign == negative => {
                current = Some((sign, trough.min(drawdown)));
            }
            Some((_, trough)) => {
                episodes.push(trough);
                current = Some((negative, drawdown));
            }
        }
    }
    episodes.push(current.unwrap().1);
    episodes.sort_by(f64::total_cmp);
    let index = (episodes.len() - 1) as f64 * (1.0 - confidence);
    let lower = index.floor() as usize;
    let upper = index.ceil() as usize;
    let weight = index - lower as f64;
    let boundary = episodes[lower] + (episodes[upper] - episodes[lower]) * weight;
    let selected: Vec<_> = episodes
        .into_iter()
        .take_while(|x| *x <= boundary)
        .collect();
    Some(-selected.iter().sum::<f64>() / selected.len() as f64)
}

#[test]
fn matches_pinned_performanceanalytics_discrete_episode_estimator() {
    let returns = [0.25, -0.20, 0.25, 0.10, -0.50, 1.0, -0.10];
    for confidence in [0.5, 0.75, 0.90, 0.95, 0.99] {
        let mut state = ConditionalDrawdownAtRisk::new(confidence, NanPolicy::Omit)
            .and_then(|mut state| {
                state.from_returns(&[])?;
                Ok(state)
            })
            .unwrap();
        let actual = state.extend(&returns).unwrap().unwrap();
        let expected = pinned_performanceanalytics_discrete(&returns, confidence).unwrap();
        assert_relative_eq!(actual, expected, epsilon = 1e-15);
    }
}

#[test]
fn includes_current_episode_and_all_boundary_ties() {
    let mut state = ConditionalDrawdownAtRisk::new(0.5, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    state.extend(&[0.25, -0.20]).unwrap();
    assert_relative_eq!(state.compute().unwrap(), 0.20, epsilon = 1e-15);
    state.append(0.25).unwrap();
    state.extend(&[0.10, -0.50]).unwrap();
    assert_relative_eq!(state.compute().unwrap(), 0.35, epsilon = 1e-15);

    let tied = [0.25, -0.20, 0.25, 0.25, -0.20];
    let mut state = ConditionalDrawdownAtRisk::new(0.95, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    assert_relative_eq!(state.extend(&tied).unwrap().unwrap(), 0.20, epsilon = 1e-15);
}

#[test]
fn bulk_scalar_chunk_reset_and_cached_compute_are_invariant() {
    let values = [0.25, f64::NAN, -0.20, 0.25, 0.10, -0.50, 1.0, -0.10];
    let mut batch = ConditionalDrawdownAtRisk::new(0.75, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    let expected = batch.extend(&values).unwrap().unwrap();
    assert_eq!(batch.len(), 7);
    assert_eq!(batch.compute(), Some(expected));
    assert_eq!(batch.compute(), Some(expected));

    let mut scalar = ConditionalDrawdownAtRisk::new(0.75, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    for value in values {
        scalar.append(value).unwrap();
    }
    assert_eq!(scalar.compute(), Some(expected));

    scalar.reset();
    assert!(scalar.is_empty());
    assert_eq!(scalar.compute(), None);
    scalar.extend(&values[..4]).unwrap();
    scalar.extend(&values[4..]).unwrap();
    assert_eq!(scalar.compute(), Some(expected));
}

#[test]
fn semantic_inputs_match_returns() {
    let returns = [0.10_f64, -0.20, 0.05];
    let logs: Vec<_> = returns.iter().map(|value| value.ln_1p()).collect();
    let equity = [100.0, 110.0, 88.0, 92.4];
    let pnl = [10.0, -22.0, 4.4];

    let compute = |kind, values: &[f64]| {
        let mut state = ConditionalDrawdownAtRisk::new(0.95, NanPolicy::Omit)
            .and_then(|mut state| {
                match kind {
                    MetricInputKind::Returns => {
                        state.from_returns(&[])?;
                    }
                    MetricInputKind::LogReturns => {
                        state.from_log_returns(&[])?;
                    }
                    MetricInputKind::Equity => {
                        state.from_equity(&[])?;
                    }
                    MetricInputKind::PeriodPnl { initial_capital } => {
                        state.from_pnl(&[], initial_capital)?;
                    }
                    MetricInputKind::RawPnl | MetricInputKind::Trades => {
                        state.append(0.0)?;
                    }
                }
                Ok(state)
            })
            .unwrap();
        state.extend(values).unwrap().unwrap()
    };
    let expected = compute(MetricInputKind::Returns, &returns);
    assert_relative_eq!(
        compute(MetricInputKind::LogReturns, &logs),
        expected,
        epsilon = 1e-15
    );
    assert_relative_eq!(
        compute(MetricInputKind::Equity, &equity),
        expected,
        epsilon = 1e-15
    );
    assert_relative_eq!(
        compute(
            MetricInputKind::PeriodPnl {
                initial_capital: 100.0,
            },
            &pnl,
        ),
        expected,
        epsilon = 1e-15
    );
}

#[test]
fn validates_configuration_domains_and_observations() {
    for confidence in [0.0, 1.0, -0.1, 1.1, f64::NAN, f64::INFINITY] {
        assert!(ConditionalDrawdownAtRisk::new(confidence, NanPolicy::Omit)
            .and_then(|mut state| {
                state.from_returns(&[])?;
                Ok(state)
            })
            .is_err());
    }
    assert!(ConditionalDrawdownAtRisk::new(0.95, NanPolicy::Omit)
        .and_then(|mut state| {
            state.append(0.0)?;
            Ok(state)
        })
        .is_err());
    assert!(ConditionalDrawdownAtRisk::new(0.95, NanPolicy::Omit)
        .and_then(|mut state| {
            state.append(0.0)?;
            Ok(state)
        })
        .is_err());

    let mut state = ConditionalDrawdownAtRisk::new(0.95, NanPolicy::Raise)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    assert!(state.append(f64::NAN).is_err());
    assert!(state.is_empty());
    assert!(state.append(f64::INFINITY).is_err());
    assert!(state.is_empty());
    assert!(state.append(-1.01).is_err());
    assert!(state.is_empty());
}
